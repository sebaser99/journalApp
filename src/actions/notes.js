import { db } from "../firebase/firebase-config";
import { addDoc, collection, doc, updateDoc, deleteDoc } from "firebase/firestore";
import { types } from "../types/types";
import { loadNotes } from "../helpers/loadNotes";
import Swal from "sweetalert2";
import { fileUpload } from "../helpers/fileUpload";

export const startNewNote = ()=>{
    return async(dispatch, getState) =>{
        const uid = getState().auth.uid;
        const newNote = {
            title: '',
            body: '',
            date: new Date().getTime(),
            url: ''
        }
        try{
            const docRef = await addDoc(collection(db, `${uid}/journal/notes`), newNote)
            console.log("Document written with ID: ", docRef.id)
            dispatch(activeNote(docRef.id, newNote))
            dispatch(addNewNote(docRef.id, newNote))
        }
        catch(e){
            console.log("Error adding a document:" , e)
        }
    }
}

export const addNewNote = (id, note)=> ({
    type: types.notesAddNew,
    payload: {
        id,
        ...note
    }
})

export const activeNote = (id, note)=>(
    {
        type: types.notesActive,
        payload: {
            id,
            ...note
        }

    }
) 

export const startLoadingNotes = (uid) =>{
    return async (dispatch) =>{
        const notes = await loadNotes(uid)
        dispatch(setNotes(notes))
    }
}

export const setNotes = (notes)=>(
    {
        type: types.notesLoad,
        payload: notes
    }
)

export const startSaveNote = (note)=>{
    return async (dispatch, getState)=> {
        const {uid} = getState().auth
        if(!note.url){
            delete note.url
        }
        const noteToSave = {...note}
        delete noteToSave.id
        await updateDoc(doc(db, `${uid}/journal/notes/${note.id}`), noteToSave)

        dispatch(refreshNoteUpdated(note.id, noteToSave))
        Swal.fire('Saved', note.title, "success")
        
    }
}

export const refreshNoteUpdated = (id, note) => ({
    type: types.notesUpdated,
    payload: {
        id, 
        note: {
            id,
            ...note
        }
    }
})

export const startUploading = (file) => {
    return async(dispatch, getState)=> {
        const {active: activeNote} = getState().notes;
        Swal.fire({
            title: 'Uploading',
            text: 'Please wait....',
            didOpen: ()=> {
                Swal.showLoading()
            }
        })
       const urlLoad = await fileUpload(file)
       activeNote.url = urlLoad

       dispatch(startSaveNote(activeNote))
       Swal.close()
    }
}

export const startDeleting = (id)=> {
    return async(dispatch, getState)=>{
        
        const uid = getState().auth.uid
        console.log(uid)
        await deleteDoc(doc(db, `${uid}/journal/notes/${id}`))
        dispatch(deleteNote(id))
    }
}

export const deleteNote = (id)=> ({
    type: types.notesDelete,
    payload: id
})

export const cleaningLogout = () => ({
    type: types.notesLogoutCleaning

})

export const searchingNotes = (search) => ({
    type: types.notesSearch,
    payload: search
}) 