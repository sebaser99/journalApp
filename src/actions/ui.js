import { types } from "../types/types";

export const uiSetError = (message)=>({
    type: types.uiSetError,
    payload: {
        msg: message
    }
})

export const uiRemoveError = ()=>({
    type: types.uiRemoveError
    
})

export const startLoading = ()=>({
    type: types.uiStartLoading,
  
})
export const finishLoading = ()=>({
    type: types.uiFinishLoading,
   
})
