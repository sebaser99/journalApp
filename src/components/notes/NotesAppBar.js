import React from 'react'

export const NotesAppBar = () => {
    return (
        <div className='notes__appbar'>
            <span>7 de enero 2022</span>
            <div>
                <button className='btn'>Picture</button>
                <button className='btn'>Save</button>
            </div>            
        </div>
    )
}
