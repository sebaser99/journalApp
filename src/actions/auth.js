import { auth } from "../firebase/firebase-config";
import { GoogleAuthProvider, signInWithPopup, createUserWithEmailAndPassword, updateProfile, signInWithEmailAndPassword } from "firebase/auth";
import { types } from "../types/types";
import { finishLoading, startLoading } from "./ui";
import Swal from 'sweetalert2';
import { cleaningLogout } from "./notes";

export const startLoginEmailPassword = (email, password)=> {
    return(dispatch)=>{
        dispatch(startLoading())

        return signInWithEmailAndPassword(auth, email, password)
            .then(({user}) =>{
                dispatch(login(user.uid, user.displayName))
                dispatch(finishLoading())
            })
            .catch(error =>{
                dispatch(finishLoading())
                Swal.fire('Error', error.message, 'error')
            }) 
        
    }
}

export const createUserWithEmailPassword = (email, password, name) =>{
    return(dispatch)=>{
         return createUserWithEmailAndPassword(auth, email, password)
        .then(async({user}) =>{
            await updateProfile(auth.currentUser, {
                displayName: name
            })
             dispatch(login(user.uid, user.displayName))
        })
        .catch(error => {
            Swal.fire('Error', error.message, 'error')
            console.log(error.code)
            
        })
    }
}

export const startGoogleLogin = ()=>{
    const googleAuthProvider = new GoogleAuthProvider();
    return(dispatch)=>{
        signInWithPopup(auth, googleAuthProvider)
        .then(({user})=>{
           dispatch(
            login(user.uid, user.displayName)
           )
  
        })
        .catch(err => console.log(err))
    }
}

export const login = (uid, displayName) => (
    {
        type: types.login,
        payload: {
            uid,
            displayName
        }
    }
)

export const startLogout = ()=> {
    return(dispatch)=>{
        auth.signOut()
        dispatch(logout())
        dispatch(cleaningLogout())
    }
}

export const logout = ()=> (
    {
        type: types.logout
    }
)

