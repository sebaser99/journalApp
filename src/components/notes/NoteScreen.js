import React from 'react'
import { NotesAppBar } from './NotesAppBar'

export const NoteScreen = () => {
    return (
        <div className='notes__main-content'>
            <NotesAppBar />

            <div className='notes__content'>
                <input type='text'
                    placeholder='Some awesome title'
                    className='notes__title-input' 
                    autoComplete='off'
                />
                <textarea
                    placeholder='What happened today'
                    className='notes__textarea'>
                </textarea>
                <div className='notes__image'>
                    <img src='https://img.vixdata.io/pd/jpg-large/es/sites/default/files/btg/universo-observable-en-una-imagen-3.png'
                    alt='image'
                    />
                    
                </div>
                

            </div>
        </div>
    )
}
