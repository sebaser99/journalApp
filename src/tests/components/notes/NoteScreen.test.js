import { mount } from "enzyme";
import React from 'react';
import { Provider } from "react-redux";
import { MemoryRouter } from "react-router-dom";

import configureStore  from 'redux-mock-store';
import thunk from 'redux-thunk';
import '@testing-library/jest-dom';
import { NoteScreen } from "../../../components/notes/NoteScreen";
import { activeNote } from "../../../actions/notes";


jest.mock('../../../actions/notes', ()=> ({
    activeNote: jest.fn()
}))



const middlewares = [thunk]
const mockStore = configureStore(middlewares)
const initState = {
    auth: {
        uid: '23',
        name: 'Sebastian'
    },
    ui: {
        loading: false,
        msgError: null
    },
    notes: {
        active:{
            id: 'ABCD',
            title: 'Hola',
            body: 'Mundo',
            date: 0,
            url: 'jgjgghghghgh'
        },
        notes: []
    }
}


let store = mockStore(initState)
store.dispatch = jest.fn()

const wrapper = mount(
    <Provider store={store}>
        <MemoryRouter>
            <NoteScreen />
        </MemoryRouter>
    </Provider>
    )

describe('vamos a probar el <NoteScreen />', ()=> {
    test('debe hacer match con el snapahot', ()=> {

        expect(wrapper).toMatchSnapshot()
    })

    test('debe hacer el llamado a activeNote', ()=> {
        wrapper.find('input[name="title"]').simulate('change', {
            target: {
                name: 'title',
                value: 'Hola dos veces'
            }
        })
        expect(activeNote).toHaveBeenLastCalledWith(
            'ABCD',
            {
                id: 'ABCD',
                title: 'Hola dos veces',
                body: 'Mundo',
                date: 0,
                url: 'jgjgghghghgh'

            }
        )
    })

})