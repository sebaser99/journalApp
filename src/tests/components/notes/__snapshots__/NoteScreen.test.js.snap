// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`vamos a probar el <NoteScreen /> debe hacer match con el snapahot 1`] = `
<div
  className="notes__main-content"
>
  <div
    className="notes__appbar"
  >
    <span>
      Wednesday, 31st December 1969
    </span>
    <input
      id="chooseFile"
      onChange={[Function]}
      style={
        Object {
          "display": "none",
        }
      }
      type="file"
    />
    <div>
      <button
        className="btn"
        onClick={[Function]}
      >
        Picture
      </button>
      <button
        className="btn"
        onClick={[Function]}
      >
        Save
      </button>
    </div>
  </div>
  <div
    className="notes__content"
  >
    <input
      autoComplete="off"
      className="notes__title-input"
      name="title"
      onChange={[Function]}
      placeholder="Some awesome title"
      type="text"
      value="Hola"
    />
    <textarea
      className="notes__textarea"
      name="body"
      onChange={[Function]}
      placeholder="What happened today"
      value="Mundo"
    />
    <div
      className="notes__image"
    >
      <img
        alt="universe"
        src="jgjgghghghgh"
      />
    </div>
    <button
      className="btn btn-danger"
      onClick={[Function]}
    >
      Delete
    </button>
  </div>
</div>
`;
